scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -524,158 +525,174 @@
                         (map job->usage)
                         (reduce (partial merge-with +)))))))
 
+(defn category->pending-jobs->category->considerable-jobs
+  "Limit the pending jobs to considerable jobs based on usage and quota.
+   Further limit the considerable jobs to a maximum of num-considerable jobs."
+  [db category->pending-jobs user->quota user->usage num-considerable]
+  (log/debug "There are" (apply + (map count category->pending-jobs)) "pending jobs")
+  (log/debug "pending-jobs:" category->pending-jobs)
+  (let [filter-considerable-jobs (fn filter-considerable-jobs [jobs]
+                                   (->> jobs
+                                        (map #(d/entity db (:db/id %)))
+                                        (filter-based-on-quota user->quota user->usage)
+                                        (filter (fn [job]
+                                                  (util/job-allowed-to-start? db job)))
+                                        (take num-considerable)))
+        category->considerable-jobs (->> category->pending-jobs
+                                         (map-vals filter-considerable-jobs))]
+    (log/debug "We'll consider scheduling" (map (fn [[k v]] [k (count v)]) category->considerable-jobs)
+               "of those pending jobs (limited to " num-considerable " due to backdown)")
+    category->considerable-jobs))
+
+(defn matches->category->job-uuids
+  "Returns the matched job uuid sets by category."
+  [matches]
+  (let [category->jobs (group-by util/categorize-job
+                                 (->> matches
+                                      (map #(-> % :tasks vec))
+                                      flatten
+                                      (map #(-> % .getRequest :job))))
+        category->job-uuids (map-vals #(set (map :job/uuid %)) category->jobs)]
+    (log/debug "matched jobs:" (map-vals count category->job-uuids))
```

Comment:
I prefer prefixing with `pc/` too. However, `map-vals` was already referred explicitly and I did not want to introduce another round of refactoring in this file. Maybe, we can do that cleanup as part of another PR.